File: DMOJ/CCC/EscapeRoom.cpp
// CCC '20 S2 - Escape Room

#include <bits/stdc++.h>
using namespace std;
// clang-format off
#ifdef __linux__
#define getchar getchar_unlocked
#endif
#define su(x) do{while((x=getchar())<48); for(x-=48; 48<=(_=getchar()); x=(x<<3)+(x<<1)+_-48);}while(0)
#define si(x) do{while((x=getchar())<45); _sign=x==45; if(_sign) while((x=getchar())<48); for(x-=48; 48<=(_=getchar()); x=(x<<3)+(x<<1)+_-48); x=_sign?-x:x;}while(0)
#define sc(x) do{while((x=getchar())<33);}while(0)
char _; bool _sign;
// clang-format on

const int MM = 1e6 + 5;
bool vis[MM];
vector<int> adj[MM];
int n, m;
int main() {
    sc(n);
    sc(m);
    deque<int> q;
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) {
            int a;
            sc(a);
            adj[i * j].push_back(a);
            if (i == 1 && j == 1) {
                q.push_back(a);
                vis[a] = true;
            }
        }
    }
    while (!q.empty()) {
        int cur = q.front();
        q.pop_front();
        if (cur == m * n) {
            printf("yes\n");
            return 0;
        }
        for (auto nxt : adj[cur]) {
            if (!vis[nxt]) {
                vis[nxt] = true;
                q.push_back(nxt);
            }
        }
    }
    printf("no\n");

    return 0;
}

File: DMOJ/CCC/EscapeRoom.py
# CCC '20 S2 - Escape Room

import sys
from collections import deque
input = sys.stdin.readline

MM = int(1e6 + 5)
vis = [False for i in range(MM)]
adj = [[] for i in range(MM)]

n = int(input())
m = int(input())

q = deque()
for i in range(1, n + 1):
    for j in range(1, m + 1):
        a = int(input())
        adj[i * j].append(a)
        if i == j == 1:
            q.append(a)
            vis[a] = True

while q:
    cur = q.pop()
    if cur == m * n:
        print("yes")
        sys.exit(0)
    for nxt in adj[cur]:
        if not vis[nxt]:
            vis[nxt] = True
            q.append(nxt)

print("no")